### PR TITLE
[FLINK-33116][tests] Fix CliClientTest.testCancelExecutionInteractiveMode fails with NPE

### PR DESCRIPTION
**Pull request by Jiabao-Sun**
## What is the purpose of the change

[FLINK-33116][tests] Fix CliClientTest.testCancelExecutionInteractiveMode fails with NPE

```shell
Sep 18 02:26:15 02:26:15.743 [ERROR] org.apache.flink.table.client.cli.CliClientTest.testCancelExecutionInteractiveMode  Time elapsed: 0.1 s  <<< ERROR!
Sep 18 02:26:15 java.lang.NullPointerException
Sep 18 02:26:15 	at org.apache.flink.table.client.cli.CliClient.closeTerminal(CliClient.java:284)
Sep 18 02:26:15 	at org.apache.flink.table.client.cli.CliClient.close(CliClient.java:108)
Sep 18 02:26:15 	at org.apache.flink.table.client.cli.CliClientTest.testCancelExecutionInteractiveMode(CliClientTest.java:314)
Sep 18 02:26:15 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
```

## Brief change log

The CliClient is not thread-safe: The close call fails with a NullPointerException because CliClient.close() calls closeTerminal() if a terminal set. But the test finishes the client.executeInInteractiveMode() in a separate thread which calls closeTerminal() in the end as well. 

Add a `CountDownLatch` to prevent that race condition.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by Jiabao-Sun** [comment]
Hi @XComp, could you please help review this?
Thanks a lot.

**Comment by Jiabao-Sun** [comment]
Hi @fsk119, could you help review this when you have time?
Thanks.

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks for looking into this, @Jiabao-Sun . I went over the change and have some comments. PTAL

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliClientTest.java [hunk 1]
```diff
@@ -300,6 +303,8 @@ void testCancelExecutionInteractiveMode() throws Exception {
                                 try {
                                     client.executeInInteractiveMode();
                                 } catch (Exception ignore) {
+                                } finally {
```
[the remark this answers is not in the document]
Comment (Jiabao-Sun) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliClientTest.java]: Thanks @XComp for the review. It makes sense to me.
I found that `CheckedThread` can solve this problem as well. 
Would you mind helping review it again?

**Review by fsk119 — APPROVED** [review]
Thanks a lot. LGTM

**Review by XComp — CHANGES_REQUESTED** [review]
Sorry for not checking it earlier, but the same issue can happen in `testCancelExecutionInNonInteractiveMode`. So, we could provide a fix to there as well.

**Comment by Jiabao-Sun** [comment]
> Sorry for not checking it earlier, but the same issue can happen in `testCancelExecutionInNonInteractiveMode`. So, we could provide a fix to there as well.

Thanks @XComp, but I don't think this issue can happen in `testCancelExecutionInNonInteractiveMode` because the `CliClient.closed()` method will be called until the newly thread was terminated. Therefore, there is no race condition in this case. WDYT?

https://github.com/apache/flink/blob/a29a320187a45ca2579a4cdeddeb99a974d9cf2d/flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliClientTest.java#L273-L279

**Review by XComp — APPROVED** [review]
Looking at the test once more, you're right. The main thread waits for the spawned thread to finish in the `while (thread.isAlive())` loop which makes it impossible for the `closeTerminal` and the `close` method to run concurrently. :+1: Thanks for double-checking my statement. I appreciate it :-)

**Comment by XComp** [comment]
Can you create backports for this fix?

**Comment by Jiabao-Sun** [comment]
> Can you create backports for this fix?

Sure.

**Comment by Jiabao-Sun** [comment]
Hi @XComp, backports is ready:
#23533 
#23534
#23535
#23536

**Review by XComp — APPROVED** [review]
Just as a hint for the future: You don't have to reassign me as a reviewer if there weren't any changes. :-)

**Pull request merged**